x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/Case.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.predicate.conditional;
+
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Expressions;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder;
+import org.elasticsearch.xpack.sql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.type.DataTypeConversion;
+import org.elasticsearch.xpack.sql.type.DataTypes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringJoiner;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+/**
+ * Implements the CASE WHEN ... THEN ... ELSE ... END expression
+ */
+public class Case extends ConditionalFunction {
+
+    private final List<IfConditional> conditions;
+    private final Expression defaultElse;
+
+    @SuppressWarnings("unchecked")
+    public Case(Source source, List<Expression> expressions) {
```

Comment:
This "weird" ctor is here as public instead of a more useful (for the `ExpressionBuilder`) one which would like:
``` public Case(Source source, List<IfConditional> conditionals, Expression defaultElse)``` because of FunctionRegistry which gets confused if there are 2 public ctors, and obviously the one with `IfConditional` cannot be used in the `FunctionRegistry`.